functions/src/api/v1/group/createGroup.ts

```diff
@@ -47,14 +64,49 @@ export const createGroup = wrap(async (req: Request, res: Response) => {
     await assertNewResourceCreationGranted(userId)
 
     const name = req.body.name
+    const requestedGroupId = req.body.groupId
+    const password = req.body.password
     const members = JSONParse(req.body.members || '[]')
 
-    const groupId = await GroupDao.add(userId, name, members)
-    await GroupDao.update({
-        id: groupId,
-    })
+    if (!password) {
+        throw new BadRequestError('Password is required')
+    }
+
+    let groupId: GroupId
+    if (requestedGroupId) {
+        groupId = await tryCreateGroupWithPredefinedId(
+            requestedGroupId,
+            userId,
+            password,
+            members
+        )
+    } else {
+        groupId = await GroupDao.add(userId, name, password, members)
+        await GroupDao.update({
+            id: groupId,
+        })
+    }
 
     res.status(201).send({
         groupId,
     })
 })
+
+const tryCreateGroupWithPredefinedId = async (
+    groupId: GroupId,
+    userId: UID,
+    password: string,
+    members: Member[]
+): Promise<GroupId> => {
+    try {
+        await GroupDao.get(groupId)
```

Comment:
missing `return groupId`

Reply:
its below, we throw directly if the await is successful.